OSDOCS#12987: Two-node OpenShift cluster with fencing (Technology Preview) (#98503)
openshift/openshift-docs: modules/installation-manual-recovering-when-auto-recovery-is-unavail.adoc

You might need to perform manual recovery steps if a disruption event prevents fencing from functioning correctly. In these cases, you can run commands directly on the control plane nodes to recover the cluster. There are four main recovery scenarios, which should be attempted in the following order:

. Update fencing secrets:  Refresh BMC credentials if they are incorrect or outdated.
. Recover a single-node failure: Restore functionality when only one control plane node is down.
. Recover a complete node failure: Restore functionality when both control plane nodes are down.
. Replace a control plane node that cannot be recovered: Replace the node to restore cluster functionality.

.Prerequisites

* You have administrative access to the control plane nodes.
* You can connect to the nodes by using SSH.

[NOTE]
====
It is recommended to take an etcd backup before proceeding. This ensures that you can restore the cluster if any issues occur.
====

.Procedure

. Update the fencing secrets:

.. If the Cluster API is unavilable, update fencing secret by running the following command on one of the cluster nodes:
+
[source,terminal]
----

> $ sudo pcs stonith update <node_name> username=<user_name> password=<password>

Reviewer: This is a small nit. Does a user know what their `node_name` is ? Also, if you just use master-0 (or master-1) it will not know about this resource. I think it is proper to use `<node_name>_redfish`.

Author: Updated as per the suggestion.

----
+
After the Cluster API recovers, or the Cluster API is already available, update fencing secret in the cluster to ensure it stays in sync, as described in the following step.

.. Edit the username and password for the existing fencing secret for the control plane node by running the following commad:
+
[source,terminal]
----
$ oc project openshift-etcd
$ oc edit secret <node_name>-fencing
----
+
If the cluster recovers after updating the fencing secrets, no further action is required. If the issue persists, proceed to the next step.

. Recover from a single-node failure:

.. Gather initial diagnostics by running the following command:
+
[source,terminal]
----
$ sudo pcs status --full
----
+

> This command provides a detailed view of the current cluster and resource states, which can help you identify issues with fencing or etcd startup.

Reviewer 2: ```suggestion
This command provides a detailed view of the current cluster and resource states. You can use to output to identify issues with fencing or etcd startup.
```

Remove anthrophomorphism

Author: Updated as per the suggestion.

.. Run additional diagnostic commands, if necessary:
+
Reset the resources on your cluster and instruct Pacemaker to attempt to start them fresh by running the following command:
+
[source,terminal]
----
$ sudo pcs resource cleanup
----
+
Review all Pacemaker activity on the node by running the following command:
+
[source,terminal]
----
$ sudo journalctl -u pacemaker
----
+
Diagnose etcd resource startup issues byrunning the following command:
+
[source,terminal]
----
$ sudo journalctl -u pacemaker | grep podman-etcd
----
+
Run these commands on a control plane node when debugging an issue with the etcd resource not starting.

.. Test fencing on the node by running the following command:
+
[source,terminal]
----

> $ sudo pcs stonith show <node_name>

Author: Updated as per the suggestion

.. Edit the username and password for the existing fencing secret for the control plane node by running the following commads:
+
[source,terminal]
----
$ oc project openshift-etcd
----
+
[source,terminal]
----
$ oc edit secret <node_name>-fencing
----
+
If the cluster recovers after updating the fencing secrets, no further action is required. If the issue persists, proceed to the next step.

. Recover from a single-node failure:

.. Gather initial diagnostics by running the following command:
+
[source,terminal]
----
$ sudo pcs status --full
----
+
This command provides a detailed view of the current cluster and resource states. You can use the output to identify issues with fencing or etcd startup.

.. Run the following additional diagnostic commands, if necessary:
+
Reset the resources on your cluster and instruct Pacemaker to attempt to start them fresh by running the following command:
+
[source,terminal]
----
$ sudo pcs resource cleanup
----
+
Review all Pacemaker activity on the node by running the following command:
+
[source,terminal]
----
$ sudo journalctl -u pacemaker
----
+

> Diagnose etcd resource startup issues byrunning the following command:

Reviewer 3: ```suggestion
Diagnose etcd resource startup issues by running the following command:
```

Reviewer 3: Fix in follow-up: https://github.com/openshift/openshift-docs/pull/100785

----
+
Pacemaker automatically identifies which nodes and resources can start. User action is not required.

.. Handle out-of-sync etcd. 
+
If one node has a more up-to-date etcd, Pacemaker attempts to fence the lagging node and start it as a learner. If this process stalls, verify the Redfish fencing endpoint and credentials by running the following command:
+
[source,terminal]
----
$ sudo pcs stonith show <node_name>
----
+
If the recovery is successful, no further action is required. If the issue persists, perform manual recovery as described in the next step.

. Manually recover from node failure:

.. Identify the node with the most up-to-date etcd or restore it from backup by running the following command:
+
[source,terminal]
----
$ ETCDCTL_API=3 etcdctl endpoint status --write-out=table
----

.. Manually isolate or power off the other node by running the following command:
+
[source,terminal]
----
$ sudo pcs stonith fence <node_name>
----

.. Force pacemaker quorum unblock by running the following command:
+
[source,terminal]
----
$ sudo pcs quorum unblock
----

.. Start podman-etcd manually on the affected node by running the following command:
+
[source,terminal]
----

> $ sudo OCF_RESKEY_CRM_meta_notify_start_resource='etcd' pcs resource debug-start etcd

Author: Updated as per the suggestion.

----
+

> When the degraded mode is active, start the second node and allow pacemaker to continue recovery. If manual recovery fails, collect a must-gather and SOS report and file a bug.

Author: Updated as per the suggestion.
